test/test_selectors.py

```diff
@@ -0,0 +1,684 @@
+
+        with self.assertTakesTime(lower=SHORT_SELECT):
+            self.assertEqual(0, len(s.select(timeout=SHORT_SELECT)))
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_timing(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        key = s.register(rd, selectors.EVENT_READ)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: wr.send(b'x'))
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(upper=LONG_SELECT):
+            ready = s.select(LONG_SELECT)
+            self.assertEqual([(key, selectors.EVENT_READ)], ready)
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_interrupt_no_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: None)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=LONG_SELECT):
+            self.assertEqual([], s.select(LONG_SELECT))
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_interrupt_with_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+        key = s.get_key(rd)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, lambda *args: wr.send(b'x'))
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(upper=LONG_SELECT):
+            self.assertEqual([(key, selectors.EVENT_READ)], s.select(LONG_SELECT))
+        self.assertEqual(rd.recv(1), b'x')
+
+    @skipUnless(hasattr(signal, "alarm"), "Platform doesn't have signal.alarm()")
+    def test_select_multiple_interrupts_with_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+        s.register(rd, selectors.EVENT_READ)
+        key = s.get_key(rd)
+
+        def second_alarm(*args):
+            wr.send(b'x')
+
+        def first_alarm(*args):
+            self.set_alarm(SHORT_SELECT)
+            signal.signal(signal.SIGALRM, second_alarm)
+
+        sigalrm_handler = signal.signal(signal.SIGALRM, first_alarm)
+        self.addCleanup(signal.signal, signal.SIGALRM, sigalrm_handler)
+        self.addCleanup(signal.alarm, 0)
+
+        # Start the timer for the interrupt.
+        self.set_alarm(SHORT_SELECT)
+
+        with self.assertTakesTime(lower=SHORT_SELECT * 2, upper=LONG_SELECT):
```

Comment:
Once again, we presumably want to *fail* this test if it takes as long as `LONG_SELECT`, but this context manager would allow that.